### PR TITLE
force and check param_names if necessary

### PR DESCRIPTION
**Pull request by picciama**
these are fixes that solve issue #102 .
All changes are made in `batchglm/models/base_glm/utils.py` in the function `parse_design`.
@davidsebfischer 
You still need to check the docstring, because I am not quite familiar with the correct syntax.
Here is the diff:
```diff
@@ -22,28 +22,33 @@ def parse_design(
     Parser for design matrices.
 
     :param design_matrix: Design matrix.
-    :param param_names:
+    :param param_names: Optional column names for design_matrix (names of individual coefficients).
     :return: Tuple containing the design matrix and the parameter names.
+    :raise ValueError: if the type of design_matrix is not understood.
+    :raise ValueError: if param_names is None when type of design_matrix is numpy.ndarray or dask.array.core.Array
+    :raise AssertionError: if length of provided param_names is not equal to number of coefficients in design_matrix
     """
+
+    params = None
     if isinstance(design_matrix, patsy.design_info.DesignMatrix):
         dmat = np.asarray(design_matrix)
         params = design_matrix.design_info.column_names
     elif isinstance(design_matrix, pd.DataFrame):
         dmat = np.asarray(design_matrix)
         params = design_matrix.columns
-    elif isinstance(design_matrix, np.ndarray):
-        dmat = design_matrix
-        params = None
     elif isinstance(design_matrix, dask.array.core.Array):
         dmat = design_matrix.compute()
-        params = None
+        params = param_names
+    elif isinstance(design_matrix, np.ndarray):
+        dmat = design_matrix
+        params = param_names
     else:
         raise ValueError("type %s not recognized" % type(design_matrix))
-
-    if param_names is not None:
-        if params is None:
-            assert len(param_names) == dmat.shape[1]
-            params = param_names
+    if params is None:
+        raise ValueError(
+            'Provide names when passing design_matrix as np.ndarray or dask.array.core.Array!')
+    assert len(params) == dmat.shape[1], "Length of provided param_names is not equal to " \
+        "number of coefficients in design_matrix."
 
     return dmat, params

```

I also changed sth. in `parse_constraints` in the same file.
Even when passing `constraint_par_names`, the function still calculated `constraint_params` from the `design_par_names` before assigning `constraint_params = constraint_par_names` anyways, if they have the correct length!
With the change applied, this unnecessary calculation doesn't happen any longer.
@davidsebfischer Please have a look at the diff to determine whether this could be a problem: 
```diff
@@ -70,15 +75,15 @@ def parse_constraints(
             constraints = constraints.compute()
         # Cannot use all parameter names if constraint matrix is not identity: Make up new ones.
         # Use variable names that can be mapped (unconstrained).
-        constraint_params = ["var_"+str(i) if np.sum(constraints[:, i] != 0) > 1
-                             else dmat_par_names[np.where(constraints[:, i] != 0)[0][0]]
-                             for i in range(constraints.shape[1])]
+        if constraint_par_names is not None:
+            assert len(constraint_params) == len(constraint_par_names)
+            constraint_params = constraint_par_names
+        else:
+            constraint_params = ["var_"+str(i) if np.sum(constraints[:, i] != 0) > 1
+                                 else dmat_par_names[np.where(constraints[:, i] != 0)[0][0]]
+                                 for i in range(constraints.shape[1])]
         assert constraints.shape[0] == dmat.shape[1], "constraint dimension mismatch"
 
-    if constraint_par_names is not None:
-        assert len(constraint_params) == len(constraint_par_names)
-        constraint_params = constraint_par_names
-
     return constraints, constraint_params

```

### CODE REVIEW

**Comment by picciama** [comment]
@davidsebfischer This should work. Do you wish a unittest prior to merging that goes thorugh all the cases? I'd like to do some cleanups

**Comment by davidsebfischer** [comment]
> @davidsebfischer This should work. Do you wish a unittest prior to merging that goes thorugh all the cases? I'd like to do some cleanups

Yes that would be great, I'll let you be the judge of that!

**Review by davidsebfischer — APPROVED** [review]

File: batchglm/models/base_glm/utils.py [hunk 1]
```diff
@@ -22,28 +22,33 @@ def parse_design(
...
     :param design_matrix: Design matrix.
-    :param param_names:
+    :param param_names: Optional column names for design_matrix (names of individual coefficients).
     :return: Tuple containing the design matrix and the parameter names.
+    :raise ValueError: if the type of design_matrix is not understood.
+    :raise ValueError: if param_names is None when type of design_matrix is numpy.ndarray or dask.array.core.Array
+    :raise AssertionError: if length of provided param_names is not equal to number of coefficients in design_matrix
     """
+
+    params = None
```
[the remark this answers is not in the document]
Comment (picciama) [reply batchglm/models/base_glm/utils.py]: done. see https://github.com/theislab/batchglm/pull/103/commits/aee24c0ac106058bf55db946b9f0ca1a1e37eddf

**Comment by picciama** [comment]
unittest is done for parse_design. Also added some more type checks and updated the docstrings / types.

**Pull request merged**